src/sdk/pynni/nni/compression/torch/builtin_quantizers.py

```diff
@@ -29,23 +42,175 @@ class QAT_Quantizer(Quantizer):
+        qmax = 1 << bits - 1
+
+        # First determine the scale.
+        scale = (rmax - rmin) / (qmax - qmin)
+
+        # Zero-point computation.
+        initial_zero_point = qmin - rmin / scale
+
+        # Now we need to nudge the zero point to be an integer
+        nudged_zero_point = 0
+        if initial_zero_point < qmin:
+            nudged_zero_point = qmin
+        elif initial_zero_point > qmax:
+            nudged_zero_point = qmax
+        else:
+            nudged_zero_point = torch.round(initial_zero_point)
+
+        op.scale = scale
+        op.zero_point = nudged_zero_point
+
+    def quantize(self, bits, op, real_val):
+        """
+        quantize real value.
+
+        Parameters
+        ----------
+        bits : int
+            quantization bits length
+        op : torch.nn.module
+            target module
+        real_val : float
+            real value to be quantized
+        Returns
+        -------
+        quantized_val : float
+        """
+        transformed_val = op.zero_point + real_val / op.scale
+        qmin = 0
+        qmax = 1 << bits - 1
+        clamped_val = torch.clamp(transformed_val, qmin, qmax)
+        quantized_val = torch.round(clamped_val)
+        return quantized_val
+
+    def dequantize(self, op, quantized_val):
+        """
+        dequantize quantized value.
+
+        Parameters
+        ----------
+        op : torch.nn.module
+            target module
+        quantized_val : float
+            quantized_val value to be dequantized
+
+        Returns
+        -------
+        real_val : float
+        """
+        real_val = op.scale * (quantized_val - op.zero_point)
+        return real_val
+
+    def quantize_weight(self, weight, config, op, **kwargs):
+        """
+        overwrite default `Quantizer` `quantize_weight` method
+        """
+        quant_bits = config.get('quant_bits', {})
+        weight_bits = quant_bits.get('weight', 8)
+        if weight_bits <= 1:
             return weight
-        a = torch.min(weight)
-        b = torch.max(weight)
-        n = pow(2, config['q_bits'])
-        scale = (b-a)/(n-1)
-        zero_point = a
-        out = torch.round((weight - zero_point)/scale)
-        out = out*scale + zero_point
-        orig_type = weight.dtype
-        return out.type(orig_type)
+        if config['quant_delay'] > self.steps:
+            return weight
+        rmin, rmax = self.fixed_range_check(weight)
+        self.update_quantization_param(weight_bits, op, rmin, rmax)
+        out = self.quantize(weight_bits, op, weight)
+        out = self.dequantize(op, out)
+        return out
+
+    def quantize_output(self, output, config, op, **kwargs):
+        """
+        overwrite default `Quantizer` `quantize_output` method
+        """
+        quant_bits = config.get('quant_bits', {})
+        output_bits = quant_bits.get('output', 8)
+        if output_bits <= 1:
+            return output
+        if config['quant_delay'] > self.steps:
```

Comment:
since we compare `quant_delay` with `self.steps`, why not change it to `quant_start_step`?

Reply:
Sure, i suppose `quant_start_step` will make much more sense.